synapse/lib/json.py

```diff
@@ -0,0 +1,191 @@
+
+    Raises:
+        synapse.exc.BadJsonText: This exception is raised when there is an error
+            deserializing the provided data.
+    '''
+    if hasattr(fp, 'mode'):
+        try:
+            with mmap.mmap(fp.fileno(), 0, prot=mmap.PROT_READ) as mm:
+                with memoryview(mm) as mv:
+                    return loads(mv)
+        except ValueError:
+            mesg = 'Cannot read empty file.'
+            raise s_exc.BadJsonText(mesg=mesg)
+    else:
+        return loads(fp.read())
+
+def dumps(obj, sort_keys=False, indent=False, default=None, asbytes=False, append_newline=False):
+    '''
+    Serialize a python object to a string.
+
+    Similar to the standard library json.dumps().
+
+    Arguments:
+        obj (object): The python object to serialize.
+        sort_keys (Optional[bool]): Sort dictionary keys. Default: False.
+        indent (Optional[bool]): Include 2 spaces of indentation. Default: False.
+        default (Optional[callable]): Callback for serializing unknown object types. Default: None.
+        asbytes (Optional[bool]): Serialized data should be returned as bytes instead of a string. Default: False.
+        append_newlines (Optional[bool]): Append a newline to the end of the serialized data. Default: False.
+
+    Returns:
+        (str | bytes): The JSON serialized python object.
+
+    Raises:
+        synapse.exc.MustBeJsonSafe: This exception is raised when a python object cannot be serialized.
+    '''
+    opts = 0
+
+    if indent:
+        opts |= orjson.OPT_INDENT_2
+
+    if sort_keys:
+        opts |= orjson.OPT_SORT_KEYS
+
+    if append_newline:
+        opts |= orjson.OPT_APPEND_NEWLINE
+
+    try:
+        ret = orjson.dumps(obj, option=opts, default=default)
+        if not asbytes:
+            ret = ret.decode()
+        return ret
+    except orjson.JSONEncodeError as exc:
+        raise s_exc.MustBeJsonSafe(mesg=exc.args[0])
+
+def dump(obj, fp, sort_keys=False, indent=False, default=None, append_newline=False):
+    '''
+    Serialize a python object to a string.
+
+    Similar to the standard library json.dump().
+
+    Note: This function attempts to automatically detect if the file was opened
+    in text or binary mode by inspecting `fp.mode`.
+
+    Arguments:
+        obj (object): The python object to serialize.
+        fp (file): The python file pointer to write the serialized data to.
+        sort_keys (Optional[bool]): Sort dictionary keys. Default: False.
+        indent (Optional[bool]): Include 2 spaces of indentation. Default: False.
+        default (Optional[callable]): Callback for serializing unknown object types. Default: None.
+        append_newlines (Optional[bool]): Append a newline to the end of the serialized data. Default: False.
+
+    Returns: None
+
+    Raises:
+        synapse.exc.MustBeJsonSafe: This exception is raised when a python object cannot be serialized.
+    '''
+    asbytes = False
+    if hasattr(fp, 'mode'):
```

Comment:
Since we get to dictate the use of these file objects, couldn't we assume they have been opened with a mode we already know about?

Reply:
Fixed.